zebra-network/src/address_book.rs

```diff
@@ -10,52 +10,44 @@ use std::{
 use chrono::{DateTime, Utc};
 use tracing::Span;
 
-use crate::{
-    constants,
-    types::{MetaAddr, PeerServices},
-};
+use crate::{constants, types::MetaAddr, PeerConnectionState};
 
 /// A database of peers, their advertised services, and information on when they
 /// were last seen.
 #[derive(Debug)]
 pub struct AddressBook {
-    by_addr: HashMap<SocketAddr, (DateTime<Utc>, PeerServices)>,
-    by_time: BTreeSet<MetaAddr>,
+    /// Each known peer address has a matching `MetaAddr`
+    by_addr: HashMap<SocketAddr, MetaAddr>,
+
+    /// The span for operations on this address book.
     span: Span,
 }
 
 #[allow(clippy::len_without_is_empty)]
 impl AddressBook {
     /// Construct an `AddressBook` with the given [`tracing::Span`].
     pub fn new(span: Span) -> AddressBook {
-        AddressBook {
+        let constructor_span = span.clone();
```

Comment:
`clone`, not move?

Reply:
You'll have to spell this one out for me

Reply:
Ah, I mean, it's not obvious to me why this requires a clone, vs handing it to this constructor which then owns it, or borrowing it? Not a blocker

Reply:
The constructor moves `span`, and `Span::enter` borrows it, so we can't use both on the same span.

Specifically:
`Span::enter` takes a reference to the span in the returned guard. The returned guard is dropped at the end of the function. `rustc` assumes that the `Drop` impl can access that reference, so it doesn't allow the move into the constructor.

Reply:
This is the conservative drop check rule, which is often more strict than necessary:
https://doc.rust-lang.org/nomicon/dropck.html